File: python/xdac.py
import zmq

class Xdac:

    def __init__(self,ip,port):

        self.max_channel=40
        self.maxC=5
        self.maxV=7.5
        self.ip=ip
        self.port=port

        context=zmq.Context()
        self.req=zmq.Socket(context,zmq.REQ)
        #self.sub=zmq.Socket(context,zmq.SUB)

        self.req.setsockopt(zmq.CONNECT_TIMEOUT,500)
        self.req.setsockopt(zmq.LINGER,0)
        self.req.setsockopt(zmq.RCVTIMEO,500)
        self.req.setsockopt(zmq.SNDTIMEO,500)

        self.req.connect("tcp://"+self.ip+":"+self.port)

        #self.sub.setsockopt(zmq.LINGER,0)
        #self.sub.setsockopt(zmq.SUBSCRIBE,b"")
        #self.sub.setsockopt(zmq.CONFLATE,True)
        #self.sub.setsockopt(zmq.RCVHWM,100)

        #self.sub.connect("tcp://"+self.ip+":5556")

    
    # ====Basic methods============

    def unlock(self,key):
        msg="GETINFO:"+key
        out=self.sendreq(msg)
        print(out)

    def lock(self):
        msg="LOCK"
        out=self.sendreq(msg)

    def shutdown(self):
        msg="EXIT"
        out=self.sendreq(msg)

    def config(self,avg_samples,v_conv_time,c_conv_time):  # Measurement configuration
        #conversion time in micro seconds.
        msg="CONFIG:"+str(avg_samples)+":"+str(v_conv_time)+":"+str(c_conv_time)
        out=self.sendreq(msg)

    def setZ(self,channel):
        msg="ZERO:"+str(channel)
        out=self.sendreq(msg)

    def setV(self,channel,V): #Set voltage in V
        msg="SETV:"+str(channel)+":"+str(V)
        out=self.sendreq(msg)

    def getV(self,channel):
        msg="MEASV:"+str(channel)
        V=self.sendreq(msg)
        return V

    
    def setC(self,channel,C): # Set current in mA
        msg="SETC:"+str(channel)+":"+str(C)
        out=self.sendreq(msg)

    def getC(self,channel):
        msg="MEASC:"+str(channel)
        C=self.sendreq(msg)
        return C
    
    def setVth(self,channel,V): #Set Voltage threshold in V
        msg="SETOVT:"+str(channel)+":"+str(V)
        out=self.sendreq(msg)

    def setCth(self,channel,C): #Set Current threshold in mA
        msg="SETOCT:"+str(channel)+":"+str(C)
        out=self.sendreq(msg)

    def setVr(self,channel,r): #Set Voltage range
        #0:0-5V  1:0-10V  2:0-20V  3:0-34V
        msg="SETR:"+str(channel)+":"+str(r)
        out=self.sendreq(msg)


    #====Derived methods=====================

    def setZ_all(self):
        
        for channel in range(1,self.max_channel+1):
            self.setZ(channel)

    def setC_all(self,C):
        
        for channel in range(1,self.max_channel+1):
            self.setC(channel,C)

    def setV_all(self,V):
        
        for channel in range(1,self.max_channel+1):
            self.setV(channel,V)

    def setVr_all(self,r):
        
        for channel in range(1,self.max_channel+1):
            self.setVr(channel,r)

    def setVth_all(self,th):
        
        for channel in range(1,self.max_channel+1):
            self.setVth(channel,th)

    def setCth_all(self,th):
        
        for channel in range(1,self.max_channel+1):
            self.setCth(channel,th)

    def setV_multi(self,channels,voltages):
        
        N=len(channels)

        for i in range(N):
            self.setV(channels[i],voltages[i])

    def setC_multi(self,channels,currents):
        
        N=len(channels)

        for i in range(N):
            self.setC(channels[i],currents[i])  

    def getV_multi(self,channels):
        
        N=len(channels)
        voltages=[0]*N
        for i in range(N):
            voltages[i]=self.getV(channels[i])

        return voltages
    
    def getC_multi(self,channels):
        
        N=len(channels)
        currents=[0]*N
        for i in range(N):
            currents[i]=self.getC(channels[i])

        return currents
    
    def CVmode(self,channels,V):
        N=len(channels)
        for i in range(N):
            self.setC(channels[i],self.maxC)
            self.setV(channels[i],V)

    def CCmode(self,channels,C):
        N=len(channels)
        for i in range(N):
            self.setV(channels[i],self.maxV)
            self.setC(channels[i],C)
    

    #===Util functions==============

    def sendreq(self,msg):
        self.req.send(msg.encode('utf-8'))
        out=self.req.recv()
        return out

File: matlab/Xdac.m
%% Class to implement functionalities for nicslab XDAC-40U-R4G8

classdef Xdac
    properties
        Key;
        Ip;
        Req;
        Sub;
        MaxCh=40;
    end

    methods
        %==General methods=====
        function obj=Xdac(ip,key)
            
            obj.Ip=ip;
            obj.Key=key;
            
            ctx = zmq.core.ctx_new();
            % Create Request Socket
            req_socket = zmq.core.socket(ctx, 'ZMQ_REQ');
            req_address = ['tcp://' ip ':5555'];
            zmq.core.connect(req_socket, req_address);
            % Create Subscribe Socket
            sub_socket = zmq.core.socket(ctx, 'ZMQ_SUB');
            sub_address = ['tcp://' XDAC_IP ':5556'];
            zmq.core.connect(sub_socket, sub_address);

            obj.Req=req_socket;
            obj.Sub=sub_socket;
        end

        function out=sendreq(obj,msg)

            msg_send = unicode2native(msg, "UTF-8");
            zmq.core.send(obj.Req, msg_send);
            
            msg_recv=zmq.core.recv(obj.Req);
            out=native2unicode(msg_recv, "UTF-8");
        end

        %==SCPI methods========
        

       function unlock(obj)
            msg="GETINFO:"+obj.Key;
            obj.sendreq(msg);
       end

       function lock(obj)
            msg="LOCK";
            obj.sendreq(msg);
       end

       function config(obj,samples,v_time,c_time)
            msg="CONFIG:"+samples+":"+v_time+":"+c_time;
            obj.sendreq(msg);
       end

       function shutdown(obj)
            msg="EXIT";
            obj.sendreq(msg);
       end

       function setZ(obj) % Set zeros to all channels
           msg="ZERO";

           for ch=1:obj.MaxCh
                obj.sendreq(msg+":"+ch);
           end
       end

       function setV(obj,ch,V)
           msg="SETV";
           obj.temp_set(msg,ch,V);
       end

       function out=getV(obj,ch)
           msg="MEASV";
           out=obj.temp_get(msg,ch);

       end

       function setC(obj,ch,C)
            msg="SETC";
            obj.temp_set(msg,ch,C);

       end

       function out=getC(obj,ch)
            msg="MEASC";
            out=obj.temp_get(msg,ch);
       end

       function setVth(obj,ch,V)
            msg="SETOVT";
            obj.temp_set(msg,ch,V);
       end

       function setCth(obj,ch,C)
            msg="SETOCT";
            obj.temp_set(msg,ch,C);

       end

       function setVr(obj,ch,r)
            msg="SETR";
            obj.temp_set(msg,ch,r);
       end

        %===Dervied methods========

        function CCmode(obj,chs,v_max,c)
            msg=["SETV","SETC"];
            obj.tempc(msg,chs,v_max,c);
        end

        function CVmode(obj,chs,v,c_max)

            msg=["SETC","SETV"];
            obj.tempc(msg,chs,c_max,v);
        end

        %==Template methods========

        function out=temp_get(obj,msg,ch) % template for get functions
           
           if (isa(ch,'double'))
                len=max(size(ch));
               if len==1
                    out=obj.sendreq(msg+":"+ch);

              elseif len>1
                   out=zeros(1,len);
                   for i=1:len
                       out(i)=obj.sendreq(msg+":"+ch(i));
                   end
               else
                   disp("Invalid input");
               end


           else
               if ch=="all"
                   out=zeros(1,obj.MaxCh);
                   for i=1:obj.MaxCh
                        out(i)=obj.sendreq(msg+":"+i);
                   end

               else
                   disp("Invalid input");
               end
           

           end

          
        end

        function out=temp_set(obj,msg,ch,var1) % template for set functions
            
           if (isa(ch,'double'))
                len=max(size(ch));

               if len==1
                    out=obj.sendreq(msg+":"+ch+":"+var1);
              elseif len>1
                   for i=1:len
                       out=obj.sendreq(msg+":"+ch(i)+":"+var1(i));
                   end
               else
                   disp("Invalid input");
               end


           else
               if ch=="all"

                   for i=1:obj.MaxCh
                        out=obj.sendreq(msg+":"+i+":"+var1);
                   end

               else
                   disp("Invalid input");
               end
           

           end
        end

        function tempc(obj,msg,ch,var1,var2) % template for cc and cv modes
            
           if (isa(ch,'double'))
                len=max(size(ch));

               if len==1
                    obj.sendreq(msg(1)+":"+ch+":"+var1);
                    obj.sendreq(msg(2)+":"+ch+":"+var2);
              elseif len>1
                   for i=1:len
                       obj.sendreq(msg(1)+":"+ch(i)+":"+var1(i));
                       obj.sendreq(msg(2)+":"+ch(i)+":"+var2(i));
                   end
               else
                   disp("Invalid input");
               end


           else
               if ch=="all"

                   for i=1:obj.MaxCh
                        obj.sendreq(msg(1)+":"+i+":"+var1);
                        obj.sendreq(msg(2)+":"+i+":"+var2);
                   end

               else
                   disp("Invalid input");
               end
           

           end
        end
    end

    
end